org-ql.el

```diff
@@ -434,11 +434,16 @@ Returns cons (INHERITED-TAGS . LOCAL-TAGS)."
                                      org-file-tags)))
                                'org-ql-nil))
            all-tags)
-      (when org-group-tags
-        (unless (eq local-tags 'org-ql-nil)
-          (setq local-tags (org-ql--expand-tag-hierarchy local-tags)))
-        (unless (eq inherited-tags 'org-ql-nil)
-          (setq inherited-tags (org-ql--expand-tag-hierarchy inherited-tags))))
+      (let ((tag-groups (and org-group-tags
+                             org-current-tag-alist
+                             (org-tag-alist-to-groups org-current-tag-alist))))
```

Comment:
IIUC this means calling `org-tag-alist-to-groups` every time this function is called on a heading whose tags are not cached.  In a file with many top-level headings (e.g. I have an inbox file with years of captured top-level headings), that would mean calling this function many times, and every time it would be with the same argument and would return the same value.

If I'm correct about that, it means that this function call probably belongs elsewhere and its result should be reused, at least per-search.

I feel like we need to reconsider the design of this part of the code altogether.  It might be that this group expansion should be done at another step of the search process.  But please see the later comments I'm about to make...

Reply:
You are correct about that but this is only going to happen if `org-current-tag-alist` is non-nil. Nonetheless, I agree with you that normally that result should be cached (and I originally thought that this was the purpose of `org-tag-groups-alist`). So maybe we can add a per-file cache for that?